activemq-client/src/main/java/org/apache/activemq/ActiveMQSession.java

```diff
@@ -1956,12 +1978,17 @@ protected void send(ActiveMQMessageProducer producer, ActiveMQDestination destin
             //Set the "JMS" header fields on the original message, see 1.1 spec section 3.4.11
             message.setJMSDeliveryMode(deliveryMode);
             long expiration = 0L;
-            if (!producer.getDisableMessageTimestamp()) {
-                long timeStamp = System.currentTimeMillis();
+            long timeStamp = System.currentTimeMillis();
+            if (timeToLive > 0) {
+                expiration = timeToLive + timeStamp;
+            }
+
+            // TODO: AMQ-8500 - update this when openwire supports JMSDeliveryTime
+            message.setJMSDeliveryTime(timeStamp);
```

Comment:
This still seems to be the 'original message' at this point, so note this will throw if it was given a foreign providers JMS 1.1 impl message object, which naturally doesnt implement the method since it doesnt exist for them (remembering back to the previous discussion around why the message set/get methods shouldnt throw UOE to avoid breaking other providers, and how a 2.0 provider can allow working with anothers existing 1.1 messages).

Reply:
I'm not following you on what the recommended approach here is. You mentioned before that the field is required and should be hydrated with the timestamp value.

- ActiveMQ producer message (and foreign v2) needs value for client after message has been processed through the producer
- ActiveMQ consumer message (and foreign v2) needs value for client after message has been received

To handle a v1 message we'd have to add reflection here and check existence of the field before assignment. That sounds very expensive for every message operation. It would not be safe to cache any of the reflection objects.

What am I missing?

Reply:
Not much. You already have to check that it is a foreign message, and are already going to the expense of converting it, this cost need only apply in those cases. Relative to e.g the round trip to the broker for any reliable messaging operation the cost is likely not that significant. If you think barffing and throwing an exception is better, I would disagree.

The consumer bit has to happen either way. It can synthesize the value from the timestamp (if present) if it doesnt have a specific differnet value.

Reply:
For the record, I try to avoid barfing exceptions whenever possible ;-) 

It was unexpected to me that this level of reflection would be the solution. I'm working on it now. That foreign message test over in http is helpful and will provide good feedback

Reply:
@gemmellr what are your thoughts on returning JMSTimestamp vs zero for the v1 use case?

getFromMessageDeliveryTime(Message fromMessage)

1. Check for the existence of the getJMSDeliveryTime method
2. If so (ie.. v2 message) return value
3. If not, return the JMSTimestamp value

Earlier we had discussed using JMSTimestamp as the fallback since it is a required field. This would align the behavior across ActiveMQ usage. It looks like qpid is returning 0 in this case. Is there a specific reason to return 0 here?

Thanks

Reply:
I should probably also note that this call will fail even for a native 5.x message, if the 1.1 API is the primary interface on the classpath, as the method 'wont exist' even though 5.x implements it. To avoid that you need to use an ActiveMQMessage impl ref to call the method rather than the Message interface.

Reply:
I could go other way on that one. Both are about as non-meaningful as each other, as they will be in the past from the recievers perspective. 0 is like the 'defaullt', ala expiration etc. Tbh I'm not sure I actually know of a reason to look at the value after receiving a message, except as part of an implementation test hehe.

I think from your 'getFromMessageDeliveryTime(Message fromMessage)' inclusion you mean during the send though. In that case, strictly speaking I guess you dont necessarily need to get the value as you are about to _set_ the value.

Reply:
Agreed that it is tough to find a meaningful use case on the receiver side. Maybe detecting clock skew? Yeah.. its narrow at best. I'll go with JMSTimestamp alignment for now and see how that feels once the tests are flushed out. I think it'll have the small benefit of making the documentation more straight-foward.